File: kotlin/src/main/kotlin/Convert Sorted Array to Binary Search Tree.kt
package o.mg

fun sortedArrayToBST(nums: IntArray): TreeNode? {
    return sortedArrayToBSTIndex(nums, 0, nums.size - 1)
}

fun sortedArrayToBSTIndex(nums: IntArray, left: Int, right: Int): TreeNode? {
    if (left > right) {
        return null
    }
    val half = (left + right) / 2
    val node = TreeNode(nums[half])
    node.left = sortedArrayToBSTIndex(nums, left, half - 1)
    node.right = sortedArrayToBSTIndex(nums, half + 1, right)
    return node
}

File: c/src/convert_sorted_array_to_binary_search_tree.c
#include <stddef.h>
#include <stdlib.h>
#include "treenode.h"

struct TreeNode *sortedArrayToBSTIndex(int *nums, int numsSize, int left, int right)
{
	if (left > right)
		return NULL;
	int half = (left + right) / 2;
	struct TreeNode *newNode = (struct TreeNode *)malloc(sizeof(struct TreeNode));
	newNode->val = half[nums];
	newNode->left = sortedArrayToBSTIndex(nums, numsSize, left, half - 1);
	newNode->right = sortedArrayToBSTIndex(nums, numsSize, half + 1, right);
	return newNode;
}

struct TreeNode *sortedArrayToBST(int *nums, int numsSize)
{
	return sortedArrayToBSTIndex(nums, numsSize, 0, numsSize - 1);
}